### ISSUE TITLE
When using "Easiest way to setup Vuls-Vulsctl", vuls scan fails.

### ISSUE THREAD

**Issue by shirousa** [future-architect/vuls#986]
# What did you do? (required. The issue will be **closed** when not provided.)

Server: CentOS 8.1
Infrastructure: Microsoft Hyper-V

Install OS with minimum configuration and install Docker.
OSを最小構成でインストールし、Docker をインストールしました。

I installed vulsctl according to "Install with vulsctl on CentOS".
"Install with vulsctl on CentOS" に従い、 vulsctl をインストールしました。
https://vuls.io/docs/en/install-with-vulsctl.html

Confirmed normal operation.
正常動作を確認しました。

```
$ ls -al ~/.ssh
合計 12
drwx------. 2 shirousa shirousa  57  5月 26 22:40 .
drwx------. 4 shirousa shirousa 141  5月 26 22:40 ..
-rw-------. 1 shirousa shirousa 2622  5月 26 22:22 id_rsa
-rw-r--r--. 1 shirousa shirousa 584  5月 26 22:22 id_rsa.pub
-rw-r--r--. 1 shirousa shirousa 346  5月 26 22:30 known_hosts
```

```
$ ./scan.sh
One Line Summary
================
trial   centos8.1.1911  649 installed, 111 updatable
```

Created a ".ssh / config" file.
".ssh/config" ファイルを作成しました。

```
$ ls -al ~/.ssh
合計 12
drwx------. 2 shirousa shirousa   71  5月 26 22:43 .
drwx------. 4 shirousa shirousa  126  5月 26 22:43 ..
-rw-------. 1 shirousa shirousa    0  5月 26 22:31 config
-rw-------. 1 shirousa shirousa 2622  5月 26 22:22 id_rsa
-rw-r--r--. 1 shirousa shirousa  584  5月 26 22:22 id_rsa.pub
-rw-r--r--. 1 shirousa shirousa  346  5月 26 22:30 known_hosts
```

vuls scan fails.
vuls scan が失敗します。

```
$ ./scan.sh
[May 26 13:44:30]  INFO [localhost] Validating config...
[May 26 13:44:30]  INFO [localhost] Detecting Server/Container OS...
[May 26 13:44:30]  INFO [localhost] Detecting OS of servers...
[May 26 13:44:30] ERROR [localhost] (1/1) Failed: trial, err: [Failed to detect OS:
    github.com/future-architect/vuls/scan.detectOS
        /go/src/github.com/future-architect/vuls/scan/serverapi.go:115
  - Unable to connect via SSH. Scan with -vvv option to print SSH debugging messages and check SSH settings. If you have never SSH to the host to be scanned, SSH to the host before scanning in order to add the HostKey. vuls@192.168.10.23 port: 22
    execResult: servername: trial
      cmd: /usr/bin/ssh -tt -o StrictHostKeyChecking=yes -o LogLevel=quiet -o ConnectionAttempts=3 -o ConnectTimeout=10 -o ControlMaster=auto -o ControlPath=/root/.vuls/controlmaster-%r-trial.%p -o Controlpersist=10m vuls@192.168.10.23 -p 22 -i /root/.ssh/id_rsa -o PasswordAuthentication=no stty cols 1000; ls /etc/debian_version
      exitstatus: 255
      stdout:
      stderr:
      err: %!s(<nil>):
    github.com/future-architect/vuls/scan.detectDebian
        /go/src/github.com/future-architect/vuls/scan/debian.go:49]
[May 26 13:44:30] ERROR [localhost] Failed to init servers. err: No scannable base servers:
    github.com/future-architect/vuls/scan.InitServers
        /go/src/github.com/future-architect/vuls/scan/serverapi.go:175
```

# What did you expect to happen?

I expect that the contents of ".ssh/config" will be used by vuls scan.
".ssh/config" の内容が、vuls scan で利用されることを期待しております。

I will use "ProxyJump" in ".ssh/config".
".ssh/config" の中で、 "ProxyJump" を利用するつもりです。
https://www.redhat.com/sysadmin/ssh-proxy-bastion-proxyjump

# What happened instead?

If I change the owner of ".ssh/config", it will work normally.
".ssh/config" の所有者変更後、正常動作を確認しております。

```
$ ls -al ~/.ssh
合計 12
drwx------. 2 shirousa shirousa   71  5月 26 22:43 .
drwx------. 4 shirousa shirousa  126  5月 26 22:43 ..
-rw-------. 1 root     root        0  5月 26 22:31 config
-rw-------. 1 shirousa shirousa 2622  5月 26 22:22 id_rsa
-rw-r--r--. 1 shirousa shirousa  584  5月 26 22:22 id_rsa.pub
-rw-r--r--. 1 shirousa shirousa  346  5月 26 22:30 known_hosts
```

```
$ ./scan.sh
One Line Summary
================
trial   centos8.1.1911  649 installed, 111 updatable
```

* Current Output

Please re-run the command using ```-debug``` and provide the output below.

```
$ ./scan.sh --debug
Using default tag: latest
latest: Pulling from vuls/vuls
Digest: sha256:a26b3e3c7e7a19a470635e7f6f91874f2c725394e2a40084c45aecd5a9855a8d
Status: Image is up to date for vuls/vuls:latest
docker.io/vuls/vuls:latest
[May 26 13:55:11]  INFO [localhost] Validating config...
[May 26 13:55:11]  INFO [localhost] Detecting Server/Container OS...
[May 26 13:55:11]  INFO [localhost] Detecting OS of servers...
[May 26 13:55:11] DEBUG [localhost] Executing... ls /etc/debian_version
[May 26 13:55:11] DEBUG [localhost] execResult: servername: trial
  cmd: /usr/bin/ssh -tt -o StrictHostKeyChecking=yes -o LogLevel=quiet -o ConnectionAttempts=3 -o ConnectTimeout=10 -o ControlMaster=auto -o ControlPath=/root/.vuls/controlmaster-%r-trial.%p -o Controlpersist=10m vuls@192.168.10.23 -p 22 -i /root/.ssh/id_rsa -o PasswordAuthentication=no stty cols 1000; ls /etc/debian_version
  exitstatus: 255
  stdout:
  stderr:
  err: %!s(<nil>)
[May 26 13:55:11] ERROR [localhost] (1/1) Failed: trial, err: [Failed to detect OS:
    github.com/future-architect/vuls/scan.detectOS
        /go/src/github.com/future-architect/vuls/scan/serverapi.go:115
  - Unable to connect via SSH. Scan with -vvv option to print SSH debugging messages and check SSH settings. If you have never SSH to the host to be scanned, SSH to the host before scanning in order to add the HostKey. vuls@192.168.10.23 port: 22
    execResult: servername: trial
      cmd: /usr/bin/ssh -tt -o StrictHostKeyChecking=yes -o LogLevel=quiet -o ConnectionAttempts=3 -o ConnectTimeout=10 -o ControlMaster=auto -o ControlPath=/root/.vuls/controlmaster-%r-trial.%p -o Controlpersist=10m vuls@192.168.10.23 -p 22 -i /root/.ssh/id_rsa -o PasswordAuthentication=no stty cols 1000; ls /etc/debian_version
      exitstatus: 255
      stdout:
      stderr:
      err: %!s(<nil>):
    github.com/future-architect/vuls/scan.detectDebian
        /go/src/github.com/future-architect/vuls/scan/debian.go:49]
[May 26 13:55:11] ERROR [localhost] Failed to init servers. err: No scannable base servers:
    github.com/future-architect/vuls/scan.InitServers
        /go/src/github.com/future-architect/vuls/scan/serverapi.go:175
```

# Steps to reproduce the behaviour

Create ".ssh/config".
".ssh/config" を作成する.

# Configuration (**MUST** fill this out):

Server: CentOS 8.1
Infrastructure: Microsoft Hyper-V
Docker version 19.03.9, build 9d988398e7

* Go version (`go version`):

* Go environment (`go env`):

* Vuls environment:

```
$ docker images
REPOSITORY               TAG                 IMAGE ID            CREATED             SIZE
vuls/vuls                latest              84de217ee124        13 hours ago        64.1MB
vuls/goval-dictionary    latest              4b9437402af7        4 days ago          23.1MB
vuls/go-cve-dictionary   latest              7f889a2a4047        4 weeks ago         25.8MB
vuls/gost                latest              d23d2f8abc19        5 weeks ago         46.3MB
hello-world              latest              bf756fb1ae65        4 months ago        13.3kB
```

To check the commit hash of HEAD
$ vuls -v

```
docker run  --rm  vuls/vuls -v
vuls v0.9.6 build-20200526_003545_2374f57
```

* config.toml:

```
$ cat vulsctl/config.toml
[servers]
[servers.trial]
host = "192.168.10.23"
port = "22"
user = "vuls"
keyPath = "/root/.ssh/id_rsa"
```

* command:

**Comment by kazuminn** [comment]
@shirousa 
Please use —ssh-confin option. 
Check https://vuls.io/docs/ja/tips.html

**Comment by shirousa** [comment]
@kazuminn

Thank you for your answer.
回答をいただきありがとうございます。

I retry it use the `—ssh-config` option.
`—ssh-config` を使用して再実行しました。

It does not improve and the error continues.
改善せず、エラーが継続します。

```
$ ./scan.sh --ssh-config
Using default tag: latest
latest: Pulling from vuls/vuls
5d20c808ce19: Already exists
d9eb74c663f4: Pull complete
5550ac0a405f: Pull complete
Digest: sha256:a4d8e42f31672b55338d6bac282054acef36416774b5941f2719618f8913822d
Status: Downloaded newer image for vuls/vuls:latest
docker.io/vuls/vuls:latest
[May 26 14:31:33]  INFO [localhost] Validating config...
[May 26 14:31:33]  INFO [localhost] Detecting Server/Container OS...
[May 26 14:31:33]  INFO [localhost] Detecting OS of servers...
[May 26 14:31:33] ERROR [localhost] (1/1) Failed: trial, err: [Failed to detect OS:
    github.com/future-architect/vuls/scan.detectOS
        /go/src/github.com/future-architect/vuls/scan/serverapi.go:115
  - Unable to connect via SSH. Scan with -vvv option to print SSH debugging messages and check SSH settings. If you have never SSH to the host to be scanned, SSH to the host before scanning in order to add the HostKey. vuls@192.168.10.23 port: 22
    execResult: servername: trial
      cmd: /usr/bin/ssh -tt vuls@192.168.10.23 -p 22 -i /root/.ssh/id_rsa -o PasswordAuthentication=no stty cols 1000; ls /etc/debian_version
      exitstatus: 255
      stdout:
      stderr: Bad owner or permissions on /root/.ssh/config

      err: %!s(<nil>):
    github.com/future-architect/vuls/scan.detectDebian
        /go/src/github.com/future-architect/vuls/scan/debian.go:49]
[May 26 14:31:33] ERROR [localhost] Failed to init servers. err: No scannable base servers:
    github.com/future-architect/vuls/scan.InitServers
        /go/src/github.com/future-architect/vuls/scan/serverapi.go:175
```

I think this phenomenon is not due to a problem in vuls itself, but rather the owner when mounting .ssh in a Docker container.
本事象は、vuls 自体の問題ではなく docker container に .ssh をマウントした際の owner に起因すると考えております。

**Comment by MaineK00n** [comment]
Hi, I'm @MaineK00n.
Here's a solution for this issue.

## TL;DR
Change `--ssh-config` option to deprecated and change config.toml to describe PATH to config.
Then, inside vuls, execute `ssh -F`.

## Survey.
As Shirousa's comment(https://github.com/future-architect/vuls/issues/986#issuecomment-635454816) says, 
this problem is caused by the fact that the docker container user and the local user are different when mounting the local `.ssh/config` in the docker container.

If there is no config, SSH will be executed successfully.
```terminal
/vuls # ls -l /root/.ssh/
Total 12
-rw------- 1 1000 1000 2602 Jun 12 06:35 id_rsa
-rw-r--r-- 1 1000 1000 568 Jun 12 06:35 id_rsa.pub
-rw-r--r-- 1 root root 175 Jun 12 07:27 known_hosts
/vuls # ssh -i /root/.ssh/id_rsa mainek00n@192. 168.11.15
Welcome to Ubuntu 20.04 LTS (GNU/Linux 5.4.0-37-generic x86_64)
 * Documentation: https://help.ubuntu.com
 * Management: https://landscape.canonical.com
 * Support: https://ubuntu.com/advantage
 * MicroK8s gets a native Windows installer and command-line integration.
     * MicroK8s gets a native Windows installer and command-line integration. https://ubuntu.com/blog/microk8s-installers-windows-and-macos
0 updates can be installed immediately.
0 of these updates are security updates.
Your Hardware Enablement Stack (HWE) is supported Until April 2025.
Last login: Fri Jun 12 16:53:28 2020 from 172.17.0. 2
```

SSH has priority over user configuration (`~/.ssh/config`).
So, if there is a problem in the permission setting, SSH will fail.
```terminal
/vuls # ls -l /root/.ssh/
total 12
-rw-rw-r--    1 1000     1000             0 Jun 12 07:54 config
-rw-------    1 1000     1000          2602 Jun 12 06:35 id_rsa
-rw-r--r--    1 1000     1000           568 Jun 12 06:35 id_rsa.pub
-rw-r--r--    1 root     root           175 Jun 12 07:27 known_hosts
/vuls # ssh -i /root/.ssh/id_rsa mainek00n@192.168.11.15
Bad owner or permissions on /root/.ssh/config
```

You can use `ssh -F` to run SSH successfully.
However, the user must be able to read config.

The `-F` option is taken from man ssh.
ssh(1) page(ref: https://linux.die.net/man/1/ssh)
```
-F configfile
Specifies an alternative per-user configuration file. If a configuration file is given on the command line, the system-wide configuration file (/etc/ssh/ssh_config) will be ignored. The default for the per-user configuration file is ~/.ssh/config. 
```

```terminal
/vuls # ssh -i /root/.ssh/id_rsa -F /root/.ssh/config mainek00n@192.168.11.15
Welcome to Ubuntu 20.04 LTS (GNU/Linux 5.4.0-37-generic x86_64)
 * Documentation:  https://help.ubuntu.com
 * Management:     https://landscape.canonical.com
 * Support:        https://ubuntu.com/advantage
 * MicroK8s gets a native Windows installer and command-line integration.
     https://ubuntu.com/blog/microk8s-installers-windows-and-macos
0 updates can be installed immediately.
0 of these updates are security updates.
Your Hardware Enablement Stack (HWE) is supported until April 2025.
Last login: Fri Jun 12 16:54:09 2020 from 172.17.0.2
```

## Policy.
vuls has an option (--ssh-config) that preferentially uses `~/.ssh/config` and `/etc/ssh/ssh_config`.
With this change, the PATH to config can be defined in server unit of config.toml. 
Execute config of specified PATH with `ssh -F`.
Then, I would like to change the command line option `--ssh-config` to deprecated.


I'm sorry for my bad English.

**Comment by MaineK00n** [comment]
A note about when this PR(https://github.com/future-architect/vuls/pull/1005) is merged.

If you mount `~/.ssh` with `ro(read only)` in Docker, the connection destination cannot be written in `known_hosts` and the following error may occur.
```terminal
vulsctl on master
$ ll ~/.ssh
-rw-r--r-- 1 mainek00n mainek00n  491 Jun 15 07:41 config
-rw------- 1 mainek00n mainek00n 2.6K Jun 13 04:52 id_rsa
-rw-r--r-- 1 mainek00n mainek00n  570 Jun 13 04:52 id_rsa.pub

vulsctl on master
$ ./scan.sh
[Jun 14 22:50:06]  INFO [localhost] Validating config...
[Jun 14 22:50:06]  INFO [localhost] Detecting Server/Container OS...
[Jun 14 22:50:06]  INFO [localhost] Detecting OS of servers...
The authenticity of host '172.17.0.2 (172.17.0.2)' can't be established.
ECDSA key fingerprint is SHA256:Z/PN9+1egSpK4vS0+2eRfq81VfAYseMt98fweGwAtxM.
Are you sure you want to continue connecting (yes/no/[fingerprint])? ye
The authenticity of host '172.17.0.2 (172.17.0.2)' can't be established.
ECDSA key fingerprint is SHA256:Z/PN9+1egSpK4vS0+2eRfq81VfAYseMt98fweGwAtxM.
Are you sure you want to continue connecting (yes/no/[fingerprint])? yes
Please type 'yes', 'no' or the fingerprint: yes
Please type 'yes', 'no' or the fingerprint: yes
Please type 'yes', 'no' or the fingerprint:
[Jun 14 22:50:16] ERROR [localhost] (1/1) Failed: name, err: [Failed to detect OS:
    github.com/future-architect/vuls/scan.detectOS
        /go/src/github.com/future-architect/vuls/scan/serverapi.go:115
  - Unable to connect via SSH. Scan with -vvv option to print SSH debugging messages and check SSH settings. If you have never SSH to the host to be scanned, SSH to the host before scanning in order to add the HostKey. root@172.17.0.2 port: 22
    execResult: servername: name
      cmd: /usr/bin/ssh -tt -F /root/.ssh/config root@172.17.0.2 -p 22 -i /root/.ssh/id_rsa -o PasswordAuthentication=no stty cols 1000; ls /etc/debian_version
      exitstatus: 255
      stdout:
      stderr: Host key verification failed.

      err: %!s(<nil>):
    github.com/future-architect/vuls/scan.detectDebian
        /go/src/github.com/future-architect/vuls/scan/debian.go:49]
[Jun 14 22:50:16] ERROR [localhost] Failed to init servers. err: No scannable base servers:
    github.com/future-architect/vuls/scan.InitServers
        /go/src/github.com/future-architect/vuls/scan/serverapi.go:175
```
or
```terminal
vulsctl on master
$ ./scan.sh
[Jun 14 23:00:37]  INFO [localhost] Validating config...
[Jun 14 23:00:37]  INFO [localhost] Detecting Server/Container OS...
[Jun 14 23:00:37]  INFO [localhost] Detecting OS of servers...
The authenticity of host '172.17.0.2 (172.17.0.2)' can't be established.
ECDSA key fingerprint is SHA256:Z/PN9+1egSpK4vS0+2eRfq81VfAYseMt98fweGwAtxM.
Are you sure you want to continue connecting (yes/no/[fingerprint])? yes
The authenticity of host '172.17.0.2 (172.17.0.2)' can't be established.
ECDSA key fingerprint is SHA256:Z/PN9+1egSpK4vS0+2eRfq81VfAYseMt98fweGwAtxM.
The authenticity of host '172.17.0.2 (172.17.0.2)' can't be established.
ECDSA key fingerprint is SHA256:Z/PN9+1egSpK4vS0+2eRfq81VfAYseMt98fweGwAtxM.
Are you sure you want to continue connecting (yes/no/[fingerprint])? yes
The authenticity of host '172.17.0.2 (172.17.0.2)' can't be established.
ECDSA key fingerprint is SHA256:Z/PN9+1egSpK4vS0+2eRfq81VfAYseMt98fweGwAtxM.
Are you sure you want to continue connecting (yes/no/[fingerprint])? yes
The authenticity of host '172.17.0.2 (172.17.0.2)' can't be established.
ECDSA key fingerprint is SHA256:Z/PN9+1egSpK4vS0+2eRfq81VfAYseMt98fweGwAtxM.
Are you sure you want to continue connecting (yes/no/[fingerprint])? yes
[Jun 14 23:05:37] ERROR [localhost] Timed out while detecting servers
[Jun 14 23:05:37] ERROR [localhost] (1/1) Timed out: name
[Jun 14 23:05:37] ERROR [localhost] Failed to init servers. err: No scannable base servers:
    github.com/future-architect/vuls/scan.InitServers
        /go/src/github.com/future-architect/vuls/scan/serverapi.go:175
```

One solution is to write the following to your ssh config file:
```
Host *
     StrictHostKeyChecking no
```

**Comment by MaineK00n** [comment]
Also note the permission of the specified ssh config file. When connecting by ssh, `ssh -F` is executed internally, but if the ssh config file specified by the user performing ssh cannot be read, the following error occurs.
```terminal
$ whoami
mainek00n

$ ll ~/.ssh
-rw------- 1 test      test       491 Jun 15 07:41 config
-rw------- 1 mainek00n mainek00n 2.6K Jun 13 04:52 id_rsa
-rw-r--r-- 1 mainek00n mainek00n  570 Jun 13 04:52 id_rsa.pub
-rw-r--r-- 1 mainek00n mainek00n  222 Jun 15 08:00 known_hosts

$ ./vuls configtest
[Jun 15 08:10:38]  INFO [localhost] Validating config...
[Jun 15 08:10:38]  INFO [localhost] Detecting Server/Container OS...
[Jun 15 08:10:38]  INFO [localhost] Detecting OS of servers...
[Jun 15 08:10:38] ERROR [localhost] (1/1) Failed: ubuntu, err: [Failed to detect OS:
    github.com/future-architect/vuls/scan.detectOS
        /home/mainek00n/github/github.com/MaineK00n/vuls/scan/serverapi.go:115
  - Unable to connect via SSH. Scan with -vvv option to print SSH debugging messages and check SSH settings. If you have never SSH to the host to be scanned, SSH to the host before scanning in order to add the HostKey. root@localhost port: 32772
    execResult: servername: ubuntu
      cmd: /usr/bin/ssh -tt -F /home/mainek00n/.ssh/config root@localhost -p 32772 -i /home/mainek00n/.ssh/id_rsa -o PasswordAuthentication=no stty cols 1000; ls /etc/debian_version
      exitstatus: 255
      stdout:
      stderr: Can't open user config file /home/mainek00n/.ssh/config: Permission denied

      err: %!s(<nil>):
    github.com/future-architect/vuls/scan.detectDebian
        /home/mainek00n/github/github.com/MaineK00n/vuls/scan/debian.go:49]
[Jun 15 08:10:38] ERROR [localhost] Failed to init servers. err: No scannable base servers:
    github.com/future-architect/vuls/scan.InitServers
        /home/mainek00n/github/github.com/MaineK00n/vuls/scan/serverapi.go:175
```

To solve this, either set the permission to 644 so that other users can read the ssh config file, or execute vuls as a user who can read the ssh config file.

FYI: ssh config file permission
According to man ssh, Permission of ssh config file should be set as follows.
```
 ~/.ssh/config
             This is the per-user configuration file.  The file format and configuration options are described in ssh_config(5).  Because of the potential for abuse, this file must have strict permissions: read/write for the user,
             and not writable by others.  It may be group-writable provided that the group in question contains only the user.
```

ref: [man ssh(web page)](https://linux.die.net/man/1/ssh)
